cstrml/payment/src/lib.rs

```diff
@@ -141,71 +141,84 @@ decl_module! {
+            if !Zero::is_zero(&real_amount) {
+                if Self::slot_pay(sorder_id.clone(), real_amount.clone()).is_ok() {
 
-            // 8. Check storage order status
-            match sorder.status {
-                OrderStatus::Success => {
-                    // 9. [DB Write] (Maybe) Transfer the amount
-                    // TODO: What if this failed several time, paid will be smaller than it should be?
-                    if T::Currency::transfer(&client, &provider, real_amount, ExistenceRequirement::AllowDeath).is_ok() {
-                        // 10. [DB Write] Update ledger
-                        <Payments<T>>::mutate(&sorder_id, |ledger| {
-                            if let Some(l) = ledger {
-                                l.paid += real_amount;
-                            }
-                        });
-                        Self::deposit_event(RawEvent::PaymentSuccess(client));
-                    }
-                },
-                _ => {}
+                } else {
+                    // TODO: Deal with failure
+                }
+            } else {
             }
+        }
+    }
 
-            Ok(())
+    /// This function will check `sorder_id` is added in `Payments` and `StorageOrders`
+    fn slot_pay(sorder_id: T::Hash, real_amount: BalanceOf<T>) -> DispatchResult {
+        // 1. Get storage order
+        let sorder = T::MarketInterface::maybe_get_sorder(&sorder_id).unwrap_or_default();
+        let client = sorder.client.clone();
+        let provider = sorder.provider.clone();
+
+        // 2. [DB Write] Unreserved 1 slot amount
+        T::Currency::unreserve(
+            &client,
+            real_amount);
+
+        // 3. [DB Write] Nice move 🥳
+        // Unreserved value will be added anyway.
+        // If the status of storage order status is `Failed`,
+        // the CRUs will be just unreserved(aka, unlocked) to client-self.
+        <Payments<T>>::mutate(&sorder_id, |ledger| {
+            if let Some(p) = ledger {
+                p.unreserved += real_amount;
+            }
+        });
+
+        // 4. Check storage order status
+        match sorder.status {
+            OrderStatus::Success => {
+                // 5. [DB Write] (Maybe) Transfer the amount
+                if T::Currency::transfer(&client, &provider, real_amount, ExistenceRequirement::AllowDeath).is_ok() {
+                    // 6. [DB Write] Update ledger
+                    <Payments<T>>::mutate(&sorder_id, |ledger| {
+                        if let Some(l) = ledger {
+                            l.paid += real_amount;
+                        }
+                    });
+                    Self::deposit_event(RawEvent::PaymentSuccess(client));
+                } else {
+                    // 7. Reserve it back
+                    // TODO: Double check this behavior since it should be a workaround. Maybe a special status is better?
+                    let _ = T::Currency::reserve(&client, real_amount);
```

Comment:
This will cause the real time longer than duration. Should we extend duration? Better to have a new status for storage order.

Reply:
We can take current strategy, do nothing and keep this `TODO`, we will explain when #138 resolved.